File: 79. Word Search.py
from typing import List


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:

        def _track(i, j, idx):
            '''
            print(i,j,idx)
            for k in is_visited:
                print(k)
            '''
            if i < 0 or j < 0 or i >= row or j >= col or idx >= w_len:
                return False
            if is_visited[i][j] or board[i][j] != word[idx]:
                return False
            if idx == w_len - 1:
                return True
            is_visited[i][j] = True
            if _track(i, j + 1, idx + 1) or _track(i, j - 1, idx + 1) or _track(i - 1, j, idx + 1) or _track(i + 1, j,
                                                                                                             idx + 1):
                return True
            is_visited[i][j] = False
            return False

        row, col, w_len = len(board), len(board[0]), len(word)
        is_visited = [[False] * col for _ in range(row)]
        for i in range(row):
            for j in range(col):
                if _track(i, j, 0):
                    return True
        return False

File: 79. Word Search.c



bool DFS(char** board, int row, int col, int r, int c, char* target, int len, int k){
    if (k == len){
        return true;
    }
    if (r<0 || r>row || c<0 || c>col || board[r][c] != target[k]){
        return false;
    }
    char tmp = board[r][c];
    board[r][c] = ' ';
    if (DFS(board, row, col, r+1, c, target, len,k+1)){
        return true;
    } 
    if(DFS(board, row, col, r-1, c, target, len, k+1)){
        return true;
    }
    if(DFS(board, row, col, r, c+1, target, len, k+1)){
        return true;
    }
    if(DFS(board, row, col, r, c-1, target, len, k+1)){
        return true;
    }
    board[r][c] = tmp;
    return false;

}


bool exist(char** board, int boardSize, int* boardColSize, char * word){
    int len = strlen(word);
    int row = boardSize -1, col = *boardColSize-1;
    for (int i=0;i<=row;i++){
        for (int j=0;j<=col;j++){
            if (DFS(board, row, col, i, j, word, len, 0)){
                return true;
            }
        }
    }
    return false;
}
